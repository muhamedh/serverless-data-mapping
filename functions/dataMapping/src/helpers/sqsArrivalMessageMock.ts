export const sqsArrivalMessage: any = {
  Records: [
    {
      messageId: "802723ba-3bbe-4794-b2bc-595f4969be88",
      receiptHandle: "dummyReceiptHandle",
      body: '{ "fileName": "product/123456789_1.xml" }',
      attributes: {
        ApproximateReceiveCount: "1",
        SentTimestamp: "1690057717056",
        SenderId: "AIDAJHIPRHEMV73VRJEBU",
        ApproximateFirstReceiveTimestamp: "1690057717062",
      },
      messageAttributes: {},
      md5OfBody: "58e83147210c41c52d83911b6fbd7b68",
      eventSource: "aws:sqs",
      eventSourceARN:
        "arn:aws:sqs:us-east-1:426643868142:entry-data-validator-sqs",
      awsRegion: "us-east-1",
    },
  ],
};
