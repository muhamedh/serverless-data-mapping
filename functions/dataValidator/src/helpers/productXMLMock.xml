<?xml version="1.0" ?>
<data>
	<HeaderData>
		<TransactionID>16459</TransactionID>
	</HeaderData>
	<ProductData>
		<ProductID>16111</ProductID>
		<ProductInformation>
			<ProductName>palaeolith demiowl Jurevis</ProductName>
			<ProductDescription>valylene seen unprosperousness unequipped coursed enthalpy emptor spuriosity cephalization sylvanry plungingly ratwa shraf balletomane nucleoidioplasma unenthusiasm Belialist regulation seneschalty tetrachromatic continentalism woodyard fleshless disculpate tachythanatous rebarbarize upstamp yesterday computable mathematic</ProductDescription>
			<ProductImages>
				<Image>image_url_0</Image>
				<Image>image_url_1</Image>
				<Image>image_url_2</Image>
			</ProductImages>
			<ProductRoundel> </ProductRoundel>
		</ProductInformation>
	</ProductData>
	<SKUData>
		<SKU>
			<SKUNumber>mI*^Luo</SKUNumber>
			<SKUName>supramental unprettiness Gaelicism genialness</SKUName>
			<SKUDescription>slay anallantoidean tensional race patinous Chavante preinvent cushionless Jezebelish zygopleural nunky thioantimoniate atelorachidia amplifier unpanel Salicornia bonebinder stylishly Chiltern salmagundi metaphosphate glutton monarchical anachromasis ungovernably pensive ocarina occipitonasal nonrustable waterwise zeist sparpiece</SKUDescription>
			<SKUPrice>
				<Amount>10.99</Amount>
				<CurrencyCode>USD</CurrencyCode>
			</SKUPrice>
			<SKUImages>
				<Image>sku_image_url_1</Image>
			</SKUImages>
			<SKUInventory>
				<InventoryStatus>InStock</InventoryStatus>
			</SKUInventory>
		</SKU>
	</SKUData>
</data>
