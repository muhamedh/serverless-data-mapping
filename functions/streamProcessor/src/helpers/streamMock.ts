export const streamMockModify: any = {
  Records: [
    {
      eventID: "7855f8fc7de325801689f4b9be097b8a",
      eventName: "MODIFY",
      eventVersion: "1.1",
      eventSource: "aws:dynamodb",
      awsRegion: "us-east-1",
      dynamodb: {
        ApproximateCreationDateTime: 1690375424,
        Keys: {
          SKUNumber: {
            S: "SKUNUMBER#",
          },
          ProductID: {
            S: "12985",
          },
        },
        NewImage: {
          SKUNumber: {
            S: "SKUNUMBER#",
          },
          data: {
            M: {
              productID: {
                S: "12985",
              },
              productInformation: {
                M: {
                  productImages: {
                    L: [
                      {
                        S: "image_url_0",
                      },
                      {
                        S: "image_url_1",
                      },
                      {
                        S: "image_url_2",
                      },
                      {
                        S: "image_url_3",
                      },
                      {
                        S: "image_url_4",
                      },
                      {
                        S: "image_url_5",
                      },
                      {
                        S: "image_url_6",
                      },
                    ],
                  },
                  productRoundel: {
                    S: " ",
                  },
                  productName: {
                    S: "andesitic mutiny deconventionalize",
                  },
                  productDescription: {
                    S: "demissionary catella serpenticidal squatterdom ananda arthrophyma objectionability behaviored morbilliform machin spary promotement tetradrachma windbracing underworkman stenocardiac righty Greenland whinchacker compotatory nonpurchase dysprosia autothermy vaticide proroyal trichologist polygyny intraglacial anthropocentric octoglot diencephalic copacetic corny hawsepipe butein",
                  },
                },
              },
            },
          },
          ProductID: {
            S: "12985",
          },
          transactionID: {
            S: "631231231231",
          },
        },
        SequenceNumber: "1046900000000012012564457",
        SizeBytes: 697,
        StreamViewType: "NEW_IMAGE",
      },
      eventSourceARN:
        "arn:aws:dynamodb:us-east-1:426643868142:table/product_db/stream/2023-07-26T12:21:14.224",
    },
  ],
};

export const streamMockInsert: any = {
  Records: [
    {
      eventID: "17ee48388d4c0a46243824bd406b33e5",
      eventName: "INSERT",
      eventVersion: "1.1",
      eventSource: "aws:dynamodb",
      awsRegion: "us-east-1",
      dynamodb: {
        ApproximateCreationDateTime: 1690375498,
        Keys: {
          SKUNumber: {
            S: "SKUNUMBER#",
          },
          ProductID: {
            S: "112312312312312312312312312312985",
          },
        },
        NewImage: {
          SKUNumber: {
            S: "SKUNUMBER#",
          },
          data: {
            M: {
              productID: {
                S: "112312312312312312312312312312985",
              },
              productInformation: {
                M: {
                  productImages: {
                    L: [
                      {
                        S: "image_url_0",
                      },
                      {
                        S: "image_url_1",
                      },
                      {
                        S: "image_url_2",
                      },
                      {
                        S: "image_url_3",
                      },
                      {
                        S: "image_url_4",
                      },
                      {
                        S: "image_url_5",
                      },
                      {
                        S: "image_url_6",
                      },
                    ],
                  },
                  productRoundel: {
                    S: " ",
                  },
                  productDescription: {
                    S: "demissionary catella serpenticidal squatterdom ananda arthrophyma objectionability behaviored morbilliform machin spary promotement tetradrachma windbracing underworkman stenocardiac righty Greenland whinchacker compotatory nonpurchase dysprosia autothermy vaticide proroyal trichologist polygyny intraglacial anthropocentric octoglot diencephalic copacetic corny hawsepipe butein",
                  },
                  productName: {
                    S: "andesitic mutiny deconventionalize",
                  },
                },
              },
            },
          },
          ProductID: {
            S: "112312312312312312312312312312985",
          },
          transactionID: {
            S: "631231231231",
          },
        },
        SequenceNumber: "1047000000000012012618473",
        SizeBytes: 781,
        StreamViewType: "NEW_IMAGE",
      },
      eventSourceARN:
        "arn:aws:dynamodb:us-east-1:426643868142:table/product_db/stream/2023-07-26T12:21:14.224",
    },
  ],
};

export const streamMockRemove = {
  Records: [
    {
      eventID: "d9842599c91cd459561a4cc8364d91a8",
      eventName: "REMOVE",
      eventVersion: "1.1",
      eventSource: "aws:dynamodb",
      awsRegion: "us-east-1",
      dynamodb: {
        ApproximateCreationDateTime: 1690375505,
        Keys: {
          SKUNumber: {
            S: "SKUNUMBER#",
          },
          ProductID: {
            S: "112312312312312312312312312312985",
          },
        },
        SequenceNumber: "1047100000000012012623177",
        SizeBytes: 61,
        StreamViewType: "NEW_IMAGE",
      },
      eventSourceARN:
        "arn:aws:dynamodb:us-east-1:426643868142:table/product_db/stream/2023-07-26T12:21:14.224",
    },
  ],
};
