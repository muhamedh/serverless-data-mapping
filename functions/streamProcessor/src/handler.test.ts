import { streamMockInsert, streamMockModify, streamMockRemove } from "./helpers/streamMock";
import { mockClient } from "aws-sdk-client-mock";

import type * as Handler from "./handler";
import {
  CloudWatchEventsClient,
  PutEventsCommand,
} from "@aws-sdk/client-cloudwatch-events";

const { handler } = jest.requireActual<typeof Handler>("./handler");
jest.setTimeout(1000000);

describe("component tests for streamProcessor lambda function", () => {
  const eventBridgeMock = mockClient(CloudWatchEventsClient);
  beforeEach(() => {
    eventBridgeMock.reset();
  });

  it("should send a statefull event to eventbridge on insert stream event", async () => {
    eventBridgeMock.on(PutEventsCommand).resolves({});
    await handler(streamMockInsert);
    expect(eventBridgeMock.call(0).args[0].input).toMatchSnapshot();
  });

  it("should send a stateless event to eventbridge on modify stream event", async () => {
    eventBridgeMock.on(PutEventsCommand).resolves({});
    await handler(streamMockModify);
    expect(eventBridgeMock.call(0).args[0].input).toMatchSnapshot();
  });

  it("should send a stateless event to eventbridge on remove stream event", async () => {
    eventBridgeMock.on(PutEventsCommand).resolves({});
    await handler(streamMockRemove);
    expect(eventBridgeMock.call(0).args[0].input).toMatchSnapshot();
  });
  
});

export {};
