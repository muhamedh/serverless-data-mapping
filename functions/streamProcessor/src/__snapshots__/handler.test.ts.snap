// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component tests for streamProcessor lambda function should send a statefull event to eventbridge on insert stream event 1`] = `
{
  "Entries": [
    {
      "Detail": ""{\\"SKUNumber\\":{\\"S\\":\\"SKUNUMBER#\\"},\\"data\\":{\\"M\\":{\\"productID\\":{\\"S\\":\\"112312312312312312312312312312985\\"},\\"productInformation\\":{\\"M\\":{\\"productImages\\":{\\"L\\":[{\\"S\\":\\"image_url_0\\"},{\\"S\\":\\"image_url_1\\"},{\\"S\\":\\"image_url_2\\"},{\\"S\\":\\"image_url_3\\"},{\\"S\\":\\"image_url_4\\"},{\\"S\\":\\"image_url_5\\"},{\\"S\\":\\"image_url_6\\"}]},\\"productRoundel\\":{\\"S\\":\\" \\"},\\"productDescription\\":{\\"S\\":\\"demissionary catella serpenticidal squatterdom ananda arthrophyma objectionability behaviored morbilliform machin spary promotement tetradrachma windbracing underworkman stenocardiac righty Greenland whinchacker compotatory nonpurchase dysprosia autothermy vaticide proroyal trichologist polygyny intraglacial anthropocentric octoglot diencephalic copacetic corny hawsepipe butein\\"},\\"productName\\":{\\"S\\":\\"andesitic mutiny deconventionalize\\"}}}}},\\"ProductID\\":{\\"S\\":\\"112312312312312312312312312312985\\"},\\"transactionID\\":{\\"S\\":\\"631231231231\\"}}"",
      "DetailType": "dataInsert",
      "EventBusName": undefined,
      "Source": "stream-processor",
    },
  ],
}
`;

exports[`component tests for streamProcessor lambda function should send a stateless event to eventbridge on modify stream event 1`] = `
{
  "Entries": [
    {
      "Detail": ""{\\"SKUNumber\\":{\\"S\\":\\"SKUNUMBER#\\"},\\"ProductID\\":{\\"S\\":\\"12985\\"}}"",
      "DetailType": "dataModify",
      "EventBusName": undefined,
      "Source": "stream-processor",
    },
  ],
}
`;

exports[`component tests for streamProcessor lambda function should send a stateless event to eventbridge on remove stream event 1`] = `
{
  "Entries": [
    {
      "Detail": ""{\\"SKUNumber\\":{\\"S\\":\\"SKUNUMBER#\\"},\\"ProductID\\":{\\"S\\":\\"112312312312312312312312312312985\\"}}"",
      "DetailType": "dataRemove",
      "EventBusName": undefined,
      "Source": "stream-processor",
    },
  ],
}
`;
